# Import Libraries

In [1]:
import numpy as np #linear algebra calculation
import pandas as pd #dataset manipulation
pd.set_option('display.max_columns', None)

# Visualization tools
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

#waring ignore
import warnings
warnings.filterwarnings('ignore')

In [2]:
print('Numpy Version:', np.__version__)
print('Pandas Version:', pd.__version__)
print('Matplotlib Version:', matplotlib.__version__)
print('Seaborn Version:', sns.__version__)

Numpy Version: 1.21.5
Pandas Version: 1.4.2
Matplotlib Version: 3.5.1
Seaborn Version: 0.11.2


# Import Dataset

In [3]:
df_raw = pd.read_csv('hotel_bookings_data.csv')

#overview
df_raw.sample(5)

hotel  is_canceled  lead_time  arrival_date_year  \
36060  Resort Hotel            0          0               2019   
67668    City Hotel            1         56               2019   
64928    City Hotel            1         56               2019   
59725    City Hotel            1        166               2018   
87158    City Hotel            0          5               2018   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
36060               July                        19                          7   
67668               July                        18                          5   
64928                May                        12                         21   
59725            January                        45                          1   
87158               June                        16                         11   

       stays_in_weekend_nights  stays_in_weekdays_nights  adults  children  \
36060                        1                         0       2       0.0   
67668                        2                         6       2       0.0   
64928                        0                         3       2       0.0   
59725                        0                         3       1       0.0   
87158                        1                         3       1       0.0   

       babies       meal                  city market_segment  \
36060       0  Breakfast    Kota Jakarta Barat      Online TA   
67668       0    No Meal  Kota Jakarta Selatan      Online TA   
64928       0  Breakfast         Kota Denpasar  Offline TA/TO   
59725       0  Breakfast         Kota Denpasar  Offline TA/TO   
87158       0  Breakfast         Kota Denpasar       Aviation   

      distribution_channel  is_repeated_guest  previous_cancellations  \
36060                TA/TO                  0                       0   
67668                TA/TO                  0                       0   
64928                TA/TO                  0                       0   
59725                TA/TO                  0                       0   
87158            Corporate                  0                       0   

       previous_bookings_not_canceled  booking_changes deposit_type  agent  \
36060                               0                0   No Deposit  240.0   
67668                               0                2   No Deposit    9.0   
64928                               0                0   Non Refund   86.0   
59725                               0                0   Non Refund  236.0   
87158                               0                0   No Deposit  153.0   

       company  days_in_waiting_list customer_type     adr  \
36060      NaN                     0      Personal   60.00   
67668      NaN                     0      Personal  132.25   
64928      NaN                     0      Personal   85.00   
59725      NaN                     0      Personal  110.00   
87158      NaN                     0      Personal   95.00   

       required_car_parking_spaces  total_of_special_requests  \
36060                            1                          2   
67668                            0                          0   
64928                            0                          0   
59725                            0                          0   
87158                            0                          0   

      reservation_status  
36060          Check-Out  
67668           Canceled  
64928           Canceled  
59725           Canceled  
87158          Check-Out

# Data Understanding

In [4]:
#dataset info
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 29 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_weekdays_nights        119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

The dataset consists of 119,390 rows and 29 columns. Every column has a proper data type

In [5]:
#missing values
df_raw.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_weekdays_nights               0
adults                                 0
children                               4
babies                                 0
meal                                   0
city                                 488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
booking_changes                        0
deposit_type                           0
agent                              16340
company                           112593
days_in_waiting_list                   0
customer_type   

In [6]:
#missing values %
cols_w_null = [x for x in df_raw.columns if df_raw[x].isnull().sum() > 0]

for i in cols_w_null:
    print(f'Column {i} : {np.round(df_raw[i].isnull().sum()*100/df_raw.shape[0], 4)}% missing values')

Column children : 0.0034% missing values
Column city : 0.4087% missing values
Column agent : 13.6862% missing values
Column company : 94.3069% missing values


The 'company' column has too many missing values (94%), consider to drop this column.

In [7]:
#duplicated rows
df_raw.duplicated().sum()

33261

In [8]:
#descriptive statistics (Numericals)
num_cols = [x for x in df_raw.columns if df_raw[x].dtypes != 'O']
cat_cols = [x for x in df_raw.columns if x not in num_cols]

df_raw[num_cols].describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2018.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2017.000000   
25%         0.000000      18.000000        2018.000000   
50%         0.000000      69.000000        2018.000000   
75%         1.000000     160.000000        2019.000000   
max         1.000000     737.000000        2019.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_weekdays_nights         adults  \
count            119390.000000             119390.000000  119390.000000   
mean                  0.927599                  2.500302       1.856403   
std                   0.998613                  1.908286       0.579261   
min                   0.000000                  0.000000       0.000000   
25%                   0.000000                  1.000000       2.000000   
50%                   1.000000                  2.000000       2.000000   
75%                   2.000000                  3.000000       2.000000   
max                  19.000000                 50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent      company  days_in_waiting_list  \
count    119390.000000  103050.000000  6797.000000         119390.000000   
mean          0.221124      86.693382   189.266735              2.321149   
std           0.652306     110.774548   131.655015             17.594721   
min           0.000000       1.000000     6.000000              0.000000   
25%           0.000000       9.000000    62.000000              0.000000   
50%           0.000000      14.000000   179.000000              0.000000   
75%           0.000000     229.000000   270.000000              0.000000   
max          21.000000     535.000000   543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000             

In [9]:
#Categorical columns unique values
for i in cat_cols:
    print(df_raw[i].value_counts())

City Hotel      79330
Resort Hotel    40060
Name: hotel, dtype: int64
October      13877
September    12661
July         11791
December     11160
June         11089
August       10939
November     10508
May           9794
April         8068
January       6794
February      6780
March         5929
Name: arrival_date_month, dtype: int64
Breakfast     92310
Dinner        14463
No Meal       10650
Undefined      1169
Full Board      798
Name: meal, dtype: int64
Kota Denpasar              48590
Kabupaten Bangka           12129
Kota Yogyakarta            10415
Kota Batu                   8568
Kabupaten Bandung           7287
                           ...  
Kabupaten Malang               1
Kabupaten Bekasi               1
Kabupaten Barito Kuala         1
Kota Palembang                 1
Kabupaten Maluku Tengah        1
Name: city, Length: 177, dtype: int64
Online TA        56477
Offline TA/TO    24219
Groups           19811
Direct           12606
Corporate         5295
Complementary      743

# Data Preprocessing

In [10]:
# Copy dataset
df = df_raw.copy()

## Handling Missing Values

In [11]:
# Filling the missing values in column 'children' with 0
df['children'].fillna(0, inplace = True)

In [12]:
# Filling the missing values in column 'city' with 'Unknown'
df['city'].fillna('Unknown', inplace = True)

In [13]:
# For now, i don't know the meaning of 'agent' column. Let's temporarily fill
#the missing values with '9999'
df['agent'].fillna(9999, inplace = True)

In [14]:
#Drop the column 'company' for now and readress it if we find it usefull in the future
df = df.drop(columns = 'company')

In [15]:
#recheck the missing values
df.isnull().sum().sum()

0

## Drop Duplicates

In [16]:
df = df.drop_duplicates()

#recheck duplicates
df.duplicated().sum()

0

## Feature Engineering

In [17]:
# Change the value 'Undefined' in 'meal' column with 'No Meal'
df['meal'] = df['meal'].apply(lambda x: x if x != 'Undefined' else 'No Meal')
df['meal'].unique()

array(['Breakfast', 'Full Board', 'Dinner', 'No Meal'], dtype=object)

In [18]:
# extract new column 'total_visitor'
df['total_visitor'] = df['adults'] + df['children'] + df['babies']

#descriptive statistics
df['total_visitor'].describe()

count    86118.000000
mean         2.029924
std          0.796341
min          0.000000
25%          2.000000
50%          2.000000
75%          2.000000
max         55.000000
Name: total_visitor, dtype: float64

In [19]:
#drop the records with total visitor = 0
df = df[df['total_visitor'] > 0]

In [20]:
#clean dataset overview
df.sample(5)

hotel  is_canceled  lead_time  arrival_date_year  \
40494     City Hotel            0         25               2017   
65508     City Hotel            1         25               2019   
38160   Resort Hotel            0         55               2019   
117052    City Hotel            0        177               2019   
62001     City Hotel            1         64               2018   

       arrival_date_month  arrival_date_week_number  \
40494           September                        31   
65508                June                        14   
38160           September                        27   
117052          September                        30   
62001            February                        53   

        arrival_date_day_of_month  stays_in_weekend_nights  \
40494                          27                        1   
65508                           4                        0   
38160                           4                        2   
117052                         27                        0   
62001                          30                        0   

        stays_in_weekdays_nights  adults  children  babies       meal  \
40494                          1       1       0.0       0  Breakfast   
65508                          3       2       2.0       0  Breakfast   
38160                          5       2       0.0       0  Breakfast   
117052                         2       2       0.0       0  Breakfast   
62001                          1       2       0.0       0  Breakfast   

                       city market_segment distribution_channel  \
40494             Kota Batu         Groups                TA/TO   
65508   Kabupaten Bangkalan      Online TA                TA/TO   
38160      Kabupaten Bangka  Complementary                TA/TO   
117052     Kabupaten Bangka  Offline TA/TO                TA/TO   
62001    Kota Jakarta Timur      Online TA                TA/TO   

        is_repeated_guest  previous_cancellations  \
40494                   0                       0   
65508                   0                       0   
38160                   0                       0   
117052                  0                       0   
62001                   0                       0   

        previous_bookings_not_canceled  booking_changes deposit_type  agent  \
40494                                0                0   No Deposit    1.0   
65508                                0                0   No Deposit    9.0   
38160                                0                0   No Deposit  531.0   
117052                               0                0   No Deposit  138.0   
62001                                0                0   No Deposit    9.0   

        days_in_waiting_list customer_type     adr  \
40494                      0        Family    0.00   
65508                      0      Personal  230.00   
38160                      0      Personal  109.65   
117052                     0      Personal   89.00   
62001                      0      Personal  131.40   

        required_car_parking_spaces  total_of_special_requests  \
40494                             0                          0   
65508                             0                          0   
38160                             0                          2   
117052                            0                          1   
62001                             0                          1   

       reservation_status  total_visitor  
40494           Check-Out            1.0  
65508            Canceled            4.0  
38160           Check-Out            2.0  
117052          Check-Out            2.0  
62001            Canceled            2.0